stl/inc/xutility

```diff
@@ -4650,83 +4650,90 @@ _FwdIt fill_n(_ExPo&&, _FwdIt _Dest, _Diff _Count_raw, const _Ty& _Val) noexcept
-// being integral is a precondition of _Value_equality_is_bitwise_equality
-// clang-format off
-template <class _Elem1, class _Elem2, _Equal_to_comparator _Pr>
-    requires (sizeof(_Elem1) == sizeof(_Elem2))
-          && _Is_nonbool_integral<_Elem1> && (!is_volatile_v<_Elem1>)
-          && _Is_nonbool_integral<_Elem2> && (!is_volatile_v<_Elem2>)
-inline constexpr bool _Equal_memcmp_is_safe_helper<_Elem1, _Elem2, _Pr> =
-    _Value_equality_is_bitwise_equality<_Elem1, _Elem2>;
-
-// Determine if we can compare pointees with memcmp
-template <class _Iter1, class _Iter2, class _Pr>
-concept _Equal_memcmp_is_safe = contiguous_iterator<_Iter1> && contiguous_iterator<_Iter2> &&
-    _Equal_memcmp_is_safe_helper<remove_const_t<remove_reference_t<iter_reference_t<_Iter1>>>,
-        remove_const_t<remove_reference_t<iter_reference_t<_Iter2>>>, _Pr>;
-// clang-format on
+// _Iterators_are_contiguous<_Iter1, _Iter2> reports whether both iterators are known to be contiguous.
+// (Without concepts, this detection is limited, which will limit when we can activate the memcmp optimization.)
+// It ignores top-level constness on the iterators.
 
-#else // ^^^ __cpp_lib_concepts / !__cpp_lib_concepts vvv
+#ifdef __cpp_lib_concepts
+// When concepts are available, we can detect arbitrary contiguous iterators.
+template <class _Iter1, class _Iter2>
+_INLINE_VAR constexpr bool _Iterators_are_contiguous = contiguous_iterator<_Iter1> //
+    && contiguous_iterator<_Iter2>;
+#else // ^^^ defined(__cpp_lib_concepts) ^^^ / vvv !defined(__cpp_lib_concepts) vvv
+// When concepts aren't available, we can detect pointers. (Iterators should be unwrapped before using this.)
+template <class _Iter1, class _Iter2>
+_INLINE_VAR constexpr bool _Iterators_are_contiguous = conjunction_v<is_pointer<_Iter1>, is_pointer<_Iter2>>;
+#endif // ^^^ !defined(__cpp_lib_concepts) ^^^
 
-// allow memcmping same-size integral non-bool non-volatile bitwise types using equal_to<>
-// being integral is a precondition of _Value_equality_is_bitwise_equality
-template <class _Elem1, class _Elem2>
-_INLINE_VAR constexpr bool _Equal_memcmp_is_safe_helper<_Elem1, _Elem2,
-    enable_if_t<sizeof(_Elem1) == sizeof(_Elem2) //
-                    && _Is_nonbool_integral<_Elem1> && _Is_nonbool_integral<_Elem2> //
-                    && !is_volatile_v<_Elem1> && !is_volatile_v<_Elem2>,
-        equal_to<>>> = _Value_equality_is_bitwise_equality<_Elem1, _Elem2>;
-
-// Determine if we can compare pointees with memcmp
-template <class _Iter1, class _Iter2, class _Pr, bool = conjunction_v<is_pointer<_Iter1>, is_pointer<_Iter2>>>
-_INLINE_VAR constexpr bool _Equal_memcmp_is_safe =
-    _Equal_memcmp_is_safe_helper<remove_const_t<remove_pointer_t<_Iter1>>, remove_const_t<remove_pointer_t<_Iter2>>,
-        _Pr>;
-
-// Arbitrary iterators cannot use memcmp
+// _Equal_memcmp_is_safe<_Iter1, _Iter2, _Pr> reports whether we can activate the memcmp optimization

```

Comment:
I am wondering whether `_Equal_memcmp_is_safe` should be shortened to `_Memcmp_is_safe`. The equal doesnt really add anything

Reply:
Although the name is awkwardly long, mentioning `equal` does serve a purpose. `memcmp` is a kind of spaceship comparison, reporting less/equal/greater, but this type trait specifically determines whether it can be used to replace equality. `unsigned int == unsigned int` can be performed by comparing the bytes. However, `unsigned int < unsigned int` can't be performed by comparing the bytes, because `unsigned int` is little-endian on our platforms.

`_Lex_compare_memcmp_classify` determines whether `memcmp` can be used for `lexicographical_compare`.